design/src/rfcs/rfc0023_serialization_and_deserialization.md

```diff
@@ -0,0 +1,90 @@
+
+When serde is unable to find a field when deserializing the data, it assigns `None` to the corresponding field.  
+It is highly unlikely that deserialization of data to fail when new fields are introduced to builder types.
+
+- When fields are removed from builder   
+Users will not be able to access that removed fields via builder, however, users should be able to access it via different datatype. (e.g. `serde_json::Map`)
+
+- feature gate for deriving `serde::Serialize` and `serde::Deserialize`
+We are going to implement them behind a feature-gate.  
+The name of the feature gate is going to be `unstable-serde-de` and `unstable-serde-ser`.
+```rust
+#[cfg_attr(feature = "unstable-serde-ser"), derive(serde::Serialize)]
+#[cfg_attr(feature = "unstable-serde-de"), derive(serde::Deserialize)]
+```
+
+This allow users to have better control of their compile time when compared to implementing them both on behind a same feature-gate.
+
+## implementing `serde::Serialize` and `serde::Deserialize` on builder types
+
+I am going to implement serde traits on existing builder types, instead of introducing new data types for serialization/deserialization.
+
+Things that I considered
+- Introduction of new data types will be more complicated; This will require an implementation of a function which converts the new data type into an corresponding data type.
+- Compile time
+Introduction of new data type should result in longer compile time than the other way around.
+This one is based on my unscientific assumption; No testing has been done to verify this.
+- Binary size
+Binary size should become bigger, however, it should be managible.
+ 
+
+## Testing
+Ideally, we want every data format to work with the sdk, however,
+- there are large number of data format out there and not every one of them are implemented for serde
+- even when serde implementations are available, some of them require programmers to pass additional pieces of data (e.g. [apache-avro](https://docs.rs/apache-avro/latest/apache_avro/) requires programmers to provide schema definition) which requires considerable investment in time to test out
+- You can't simply map rust lang's data in some cases e.g. XML, CSV.  
+  To clearify, it is technicalliy possible to implement something that works with XML and CSV, but there are things to figure out, for example
+    - in case of CSV, you need to figure out how to handle nested structures
+    - in case of XML, you need to figure out what to do with attributes, child tags, contents ... etc
+
+Having said, I suggest starting with following data format.
+- JSON
+- YAML
+- TOML
+  
+Those data formats are popular and it puts strong emphasis on human-readability.  
+
+Since we could use serialized data used in testing for examples for new users, I think this is a fair start.  
+Also, many popular data format should work if it is going to work with these 3. (I have no scientific basis for this. It's just my gut feeling.)
```

Comment:
I think this statement is too vague. It would be more useful to have some specifics here. Why would other data formats work?

Reply:
I think we want to replace this section entirely with something like this.
```
Having said that, we will create tests to ensure that serialization/deserialization will always work the JSON format.  
AWS has services that makes extensive use of JSON such as JSON policy document, and we believe that many AWS users are familiar with JSON format already and even when a user is not familiar with it, it will not be difficult for a user to find a person for assistance.
  
```

Reply:
That definitely sounds better.

Thank you!